src/Components/Blazor/WebAssembly.Authentication/src/Interop/AuthenticationService.ts

```diff
@@ -0,0 +1,319 @@
+                    { scope: request.scopes.join(' ') } : undefined;
+
+                const newUser = await this._userManager.signinSilent(parameters);
+
+                return {
+                    status: AccessTokenResultStatus.Success,
+                    token: {
+                        grantedScopes: newUser.scopes,
+                        expires: getExpiration(newUser.expires_in),
+                        value: newUser.access_token
+                    }
+                };
+
+            } catch (e) {
+                return {
+                    status: AccessTokenResultStatus.RequiresRedirect
+                };
+            }
+        }
+
+        function hasValidAccessToken(user: User | null): user is User {
+            return !!(user && user.access_token && !user.expired && user.scopes);
+        }
+
+        function getExpiration(expiresIn: number) {
+            const now = new Date();
+            now.setTime(now.getTime() + expiresIn * 1000);
+            return now;
+        }
+
+        function hasAllScopes(request: AccessTokenRequestOptions | undefined, currentScopes: string[]) {
+            const set = new Set(currentScopes);
+            if (request && request.scopes) {
+                for (let current of request.scopes) {
+                    if (!set.has(current)) {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+    }
+
+    async signIn(state: any) {
+        try {
+            await this._userManager.clearStaleState();
+            const silentUser = await this._userManager.signinSilent(this.createArguments());
+            return this.success(state);
+        } catch (silentError) {
+            await this._userManager.clearStaleState();
+            // User might not be authenticated, fallback to popup authentication
+            console.log("Silent authentication error: ", silentError);
+
+            try {
+                await this._userManager.signinRedirect(this.createArguments(state));
+                return this.redirect();
+            } catch (redirectError) {
+                console.log("Redirect authentication error: ", redirectError);
+                return this.error(redirectError);
+            }
+        }
+    }
+
+    async completeSignIn(url: string) {
+        const requiresLogin = await this.loginRequired(url);
+        const stateExists = await this.stateExists(url);
+        try {
+            const user = await this._userManager.signinCallback(url);
+            if (window.self !== window.top) {
+                return this.operationCompleted();
+            } else {
+                return this.success(user && user.state);
+            }
+        } catch (error) {
+            if (requiresLogin || window.self !== window.top || !stateExists) {
+                return this.operationCompleted();
+            }
+
+            console.log('There was an error signing in: ', error);
```

Comment:
The duplication of log messages is pretty suspect. Should we be dropping L169 here and just rely on L170?

Reply:
Went away.